File: exact.py
import functions
import networkx as nx
import matplotlib.pyplot as plt
import sys
from itertools import combinations_with_replacement
from collections import Counter

def run(G, s, b, t):

    G_prime = G.copy()
    A = set()  # the set of (increased edge, delta) pair
    
    coreness = {}
    s_core_num = functions.calculate_s_core(G_prime, G_prime.nodes, s, coreness)  # Calculate s-core and coreness

    # Filter candidate edges
    non_s_core = []
    s_cand = None
    for n, d in G_prime.nodes(data=True):
        if not d['label']:
            non_s_core.append(n)
        else:
            if not s_cand:
                s_cand = n
    if s_cand is None:
        print("No node in s-core. Change s value")
        sys.exit(1)

    # generate candidate edge
    candidate_edges = []
    # 1) Between non-core nodes
    for i in range(len(non_s_core)):
        for j in range(i+1, len(non_s_core)):
            candidate_edges.append((non_s_core[i], non_s_core[j]))
    # 2) Between non-core nodes and s-core nodes
    for u in non_s_core:
        candidate_edges.append((u, s_cand))

    C = list(combinations_with_replacement(candidate_edges, b))

    best_gain = -1
    best_combo = None

    # Iterate over all combinations
    for combo in C:
        combo_map = Counter(combo)  # e.g. Counter({(u,v):2, (x,y):1, ...})
        # Create a copy for s-core calculation
        G_temp = G.copy()
        for (u,v), delta in combo_map.items():
            if G_temp.has_edge(u, v):
                G_temp[u][v]['weight'] += delta
            else:
                G_temp.add_edge(u, v, weight=delta)

        # Re-calculate s-core on the modified G_temp
        temp_coreness = {}
        new_size = functions.calculate_s_core(G_temp, G_temp.nodes, s, temp_coreness)

        gain = new_size - s_core_num
        if gain > best_gain:
            best_gain  = gain
            best_combo = combo_map

    # Apply the best combination
    for (u, v), delta in best_combo.items():
        if G_prime.has_edge(u, v):
            G_prime[u][v]['weight'] += delta
        else:
            G_prime.add_edge(u, v, weight=delta)
        A.add(((u, v), delta))

    # Final s-core size
    final_coreness = {}
    final_size = functions.calculate_s_core(G_prime, G_prime.nodes, s, final_coreness)

    return A, final_size, final_size-s_core_num

File: code/exact.cpp
#include "exact.h"
#include "functions.h"
#include <algorithm>
#include <vector>
#include <map>

// -------------------------------------------------------------------------
// Generate all combinations with replacement of size b from candidate_edges
// This is exponential; only feasible for small b and small candidate sets.
// -------------------------------------------------------------------------
static void combinations_with_replacement_helper(
    const std::vector<std::pair<int,int>>& pool, int r, int start,
    std::vector<int>& current,
    std::vector<std::vector<int>>& result) {
    if ((int)current.size() == r) {
        result.push_back(current);
        return;
    }
    for (int i = start; i < (int)pool.size(); i++) {
        current.push_back(i);
        combinations_with_replacement_helper(pool, r, i, current, result);
        current.pop_back();
    }
}

ExactResult exact_run(Graph& G, int s, int b) {
    ExactResult result;

    Graph G_prime = G.copy();
    std::vector<int> all_n;
    for (int i = 0; i < G_prime.n; i++) all_n.push_back(i);

    Coreness coreness(G_prime.n, {-1,-1});
    int s_core_num = calculate_s_core(G_prime, all_n, s, coreness);
    int initial_size = s_core_num;

    // Find non-s-core and s_cand
    std::vector<int> non_s_core;
    int s_cand = -1;
    for (int nd = 0; nd < G_prime.n; nd++) {
        if (!G_prime.label[nd]) {
            non_s_core.push_back(nd);
        } else {
            if (s_cand == -1) s_cand = nd;
        }
    }
    if (s_cand == -1) {
        result.final_size = s_core_num;
        result.followers = 0;
        return result;
    }

    // Generate candidate edges
    std::vector<std::pair<int,int>> candidate_edges;
    int nn = (int)non_s_core.size();
    for (int i = 0; i < nn; i++) {
        for (int j = i+1; j < nn; j++) {
            candidate_edges.push_back({non_s_core[i], non_s_core[j]});
        }
    }
    for (int u : non_s_core) {
        candidate_edges.push_back({u, s_cand});
    }

    // Generate all combinations_with_replacement of size b
    std::vector<std::vector<int>> combos;
    std::vector<int> cur;
    combinations_with_replacement_helper(candidate_edges, b, 0, cur, combos);

    int best_gain = -1;
    std::map<std::pair<int,int>, int> best_combo;

    for (auto& combo_indices : combos) {
        // Count occurrences
        std::map<std::pair<int,int>, int> combo_map;
        for (int idx : combo_indices) {
            combo_map[candidate_edges[idx]]++;
        }

        // Build temp graph
        Graph G_temp = G.copy();
        for (auto& [e, delta] : combo_map) {
            auto [u, v] = e;
            if (G_temp.has_edge(u, v)) {
                G_temp.increase_weight(u, v, delta);
            } else {
                G_temp.add_edge(u, v, delta);
            }
        }

        Coreness temp_cor(G_temp.n, {-1,-1});
        std::vector<int> an;
        for (int i = 0; i < G_temp.n; i++) an.push_back(i);
        int new_size = calculate_s_core(G_temp, an, s, temp_cor);
        int gain = new_size - initial_size;

        if (gain > best_gain) {
            best_gain = gain;
            best_combo = combo_map;
        }
    }

    // Apply best combo
    if (!best_combo.empty()) {
        for (auto& [e, delta] : best_combo) {
            auto [u, v] = e;
            if (G_prime.has_edge(u, v)) {
                G_prime.increase_weight(u, v, delta);
            } else {
                G_prime.add_edge(u, v, delta);
            }
            result.A.push_back({e, delta});
        }
    }

    Coreness final_cor(G_prime.n, {-1,-1});
    std::vector<int> an2;
    for (int i = 0; i < G_prime.n; i++) an2.push_back(i);
    result.final_size = calculate_s_core(G_prime, an2, s, final_cor);
    result.followers = result.final_size - initial_size;

    return result;
}
